### PR TITLE
Added pathlib support for several functions

### LINKED ISSUE #6418: Allow pathlib.Path arguments
`savez`, `savez_compressed`, and `load` all take `str` as an argument, but give an error when passed a `pathlib.Path` argument as a filename. Now that `pathlib.Path` is in the standard library, it would be nice if these functions could accept a `Path` instance as an argument, in addition to `str` and `file` objects.

Example:

```python
p = Path('test.npz')
np.savez_compressed(p, data=[10])
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-16-304bca81cdf4> in <module>()
      1 p = Path('test.npz')
----> 2 np.savez_compressed(p, data=[10])

/usr/lib/python3.5/site-packages/numpy/lib/npyio.py in savez_compressed(file, *args, **kwds)
    510 
    511     """
--> 512     _savez(file, args, kwds, True)
    513 
    514 

/usr/lib/python3.5/site-packages/numpy/lib/npyio.py in _savez(file, args, kwds, compress)
    550                 fid.close()
    551                 fid = None
--> 552                 zipf.write(tmpfile, arcname=fname)
    553             finally:
    554                 if fid:

/usr/lib/python3.5/zipfile.py in write(self, filename, arcname, compress_type)
   1481                 zip64 = self._allowZip64 and \
   1482                     zinfo.file_size * 1.05 > ZIP64_LIMIT
-> 1483                 self.fp.write(zinfo.FileHeader(zip64))
   1484                 file_size = 0
   1485                 while 1:

/usr/lib/python3.5/zipfile.py in write(self, data)
    675 
    676     def write(self, data):
--> 677         n = self.fp.write(data)
    678         self.offset += n
    679         return n

AttributeError: 'PosixPath' object has no attribute 'write'
```

### PR DESCRIPTION
**Pull request by wackywendell**
Added support for pathlib.Path support for several functions, including `savetxt`, `loadtxt`, `load`, `savez`, `savez_compressed`, `ndfromtxt`, `recfromtxt`, and `recfromcsv`, along with testing. Fixes #6418.

Note that #6655 also addresses the same issue, and we should coordinate on this; I was a bit slow on making my changes, and @r0fls did some good work on his branch!

This could also use a bit of code review, I imagine.

CC @rgommers, @r0fls.

### CODE REVIEW

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -425,7 +429,7 @@ def save(file, arr, allow_pickle=True, fix_imports=True):
 
     Parameters
     ----------
-    file : file or str
+    file : file, str, or pathlib.Path
         File or filename to which the data is saved.  If file is a file-object,
         then the filename is unchanged.  If file is a string, a ``.npy``
```
Comment (r0fls) [inline numpy/lib/npyio.py]: You could update this to say `If file is a string or a pathlib.Path object, a ''.npy''... extension will be appended to the file name if it does not already...`

Comment (wackywendell) [inline numpy/lib/npyio.py]: Good point, and done!

File: numpy/lib/npyio.py [hunk 2]
```diff
@@ -86,10 +86,12 @@ def __dir__(self):
         return object.__getattribute__(self, '_obj').keys()
 
 
-def zipfile_factory(*args, **kwargs):
+def zipfile_factory(file, mode='r', *args, **kwargs):
```
Comment (charris) [inline numpy/lib/npyio.py]: Umm, adding a positional argument seems like a potential compatibility problem.

File: numpy/lib/_iotools.py [hunk 3]
```diff
@@ -50,6 +54,13 @@ def _is_bytes_like(obj):
     return True
 
 
+def _is_path(obj):
+    """
+    Check whether obj is a pathlib.Path object.
+    """
+    return PurePath is not None and isinstance(obj, PurePath)
```
Comment (charris) [inline numpy/lib/_iotools.py]: Do you need to check `PurePath is not None`?

File: numpy/lib/npyio.py (hunk 2 above)
Comment (wackywendell) [inline numpy/lib/npyio.py]: I'm not sure how in this case it would be a problem. Could you be more specific?

Before, it was just passing `*args` to `zipfile.Zipfile()`; the first two args to  `zipfile.Zipfile()` are necessarily `file` and `mode`, and AFAICT always have been. I'm not sure how this could be a backwards-compatibility problem.

Or are you thinking this might be a *future* compatibility problem?

File: numpy/lib/_iotools.py [hunk 4]
```diff
@@ -50,6 +54,13 @@ def _is_bytes_like(obj):
     return True
 
 
+def _is_path(obj):
```
Comment (charris) [inline numpy/lib/_iotools.py]: `is_path` seems too generic for a specific type. Maybe `is_pathlib_path`.

File: numpy/lib/_iotools.py (hunk 3 above)
Comment (charris) [inline numpy/lib/_iotools.py]: `PurePath` looks like the wrong type to check as it may be purely abstract, `Path` would seem better.

File: numpy/lib/_iotools.py [hunk 5]
```diff
@@ -9,6 +9,10 @@
 import numpy as np
 import numpy.core.numeric as nx
 from numpy.compat import asbytes, bytes, asbytes_nested, basestring
+try:
+    from pathlib import Path, PurePath
+except ImportError:
+    PurePath = Path = None
```
Comment (charris) [inline numpy/lib/_iotools.py]: Do you really need `PurePath`?

File: numpy/lib/_iotools.py (hunk 3 above)
Comment (wackywendell) [inline numpy/lib/_iotools.py]: Hmm, that is an interesting thought, but I just plugged it into the interpreter and it gives a `TypeError` for `isinstance(x, None)`:

```python
>>> p=Path('x')
>>> isinstance(p, None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isinstance() arg 2 must be a type or tuple of types
```

Note that `PurePath` will be `None` if pathlib is not installed, so the `PurePath is None` is protecting that `isinstance`.

File: numpy/lib/_iotools.py (hunk 4 above)
Comment (wackywendell) [inline numpy/lib/_iotools.py]: Good point. I'll do that.

File: numpy/lib/_iotools.py (hunk 3 above)
Comment (wackywendell) [inline numpy/lib/_iotools.py]: > `PurePath` looks like the wrong type to check as it may be purely abstract, `Path` would seem better.

Ah, I misread the docs and thought `Path` was an alias for `PosixPath` or `WindowsPath`. You're right, those both inherit from `Path`.

However, it doesn't look to me like `PurePath` is ever abstract, and I'm not sure it matters anyway: I'm only using it for `isinstance`, so its fine if its abstract.

That said, there's no reason to not just go with `Path`, so maybe we should do that.

**Comment by charris** [comment]
See also #6655, which does much the same.

**Comment by wackywendell** [comment]
> See also #6655, which does much the same.

Yes, #6655 adds `pathlib.Path` support to one function; this adds it to about 7. I could rebase this onto #6655, though, if you'd like... I was waiting to see if #6655 would be merged first, though.

**Comment by charris** [comment]
The PR list tends to act like a lifo :) I just saw #6655 as I was going down the list.

File: numpy/lib/npyio.py (hunk 2 above)
Comment (charris) [inline numpy/lib/npyio.py]: By and large, signatures should not be changed without good reason, there is always the possibility that you have missed something nor can you rely on the tests being good enough to catch problems. This is particularly true for public functions, which unfortunately this is. The original was perhaps unfortunate, but there you go. It is best to keep PRs focused. If you feel this change is a useful cleanup, then it should come in as a separate PR.

Comment (charris) [inline numpy/lib/npyio.py]: Just to clarify the procedure ;) A new private function, say `_zipfile_factory` would be defined, and `zipfile_factory` would be deprecated. That may seem overly picky, but all such changes carry some risk, and when several hundred PRs are present in a release the risk accumulates.

Comment (wackywendell) [inline numpy/lib/npyio.py]: OK, that makes sense. I'll work on it - probably by changing it back, because I don't think this is important enough to warrant a deprecation cycle.

Comment (wackywendell) [inline numpy/lib/npyio.py]: I reverted the type signature, although I did add an `assert`. With the previous version, a 0-length `*args` would cause an error from `zipfile.Zipfile`; this just changes the exception that would be raised from a `TypeError` to an `AssertionError`. Is that ok, or should I also preserve the assertion type?

File: numpy/lib/npyio.py [hunk 6]
```diff
@@ -87,6 +87,13 @@ def __dir__(self):
 
 
 def zipfile_factory(*args, **kwargs):
+    """Create a zipfile.Zipfile. Allows for Zip64, as well as pathlib.Path 
+    objects."""
+    assert len(args) >= 1, "zipfile_factory requires a file"
```
Comment (charris) [inline numpy/lib/npyio.py]: I think you can keep the file argument, it isn't optional in the ZipFile constructor and you would get an error anyway without it.
```
class zipfile.ZipFile(file[, mode[, compression[, allowZip64]]])
```

Comment (charris) [inline numpy/lib/npyio.py]: Note that `assert` doesn't work when python is run in optimized mode. Just raise an error if you want to do it that way. Personally, I think a `TypeError` from `ZipFile` is fine, especially as it is unlikely to ever be seen.

Comment (wackywendell) [inline numpy/lib/npyio.py]: Alright, now its `zipfile_factory(file, *args, **kwargs)`, and the code is cleaner.

File: numpy/lib/npyio.py [hunk 7]
```diff
@@ -476,6 +482,11 @@ def save(file, arr, allow_pickle=True, fix_imports=True):
             file = file + '.npy'
         fid = open(file, "wb")
         own_fid = True
+    elif _is_pathlib_path(file):
+        if not file.name.endswith('.npy'):
+            file = file.parent / (file.name + '.npy')
```
Comment (charris) [inline numpy/lib/npyio.py]: Out of curiosity, what does the `/` do here. I haven't seen that before.

Comment (charris) [inline numpy/lib/npyio.py]: Ah, looked it up.

File: numpy/lib/tests/test_io.py [hunk 8]
```diff
@@ -1823,6 +1828,123 @@ def test_auto_dtype_largeint(self):
         assert_equal(test['f1'], 17179869184)
         assert_equal(test['f2'], 1024)
 
+
+class TestPathUsage(TestCase):
+    @contextlib.contextmanager
```
Comment (charris) [inline numpy/lib/tests/test_io.py]: This may not handle exceptions correctly, writing a standard context manager might be better.

Comment (charris) [inline numpy/lib/tests/test_io.py]: There is another PR for adding a context manager for temporary files, see #6545 and comments there.

File: numpy/lib/tests/test_io.py [hunk 9]
```diff
@@ -1823,6 +1828,123 @@ def test_auto_dtype_largeint(self):
...
+        # Used as in TestSavezLoad.test_closing_zipfile_after_load
+        
+        with tempdir(prefix=prefix) as tmpdir:
+            fd, tmp = mkstemp(dir=tmpdir, *args, **kwargs)
+            os.close(fd)
+            yield Path(tmp)
+    
+    # Test that pathlib.Path can be used
+    @np.testing.dec.skipif(Path is None,
+    "Cannot test without a pathlib.Path instance")
```
Comment (charris) [inline numpy/lib/tests/test_io.py]: Indent should line up with the `P` in `Path` above. See PEP8.

File: numpy/lib/tests/test_io.py [hunk 10]
```diff
@@ -1823,6 +1828,123 @@ def test_auto_dtype_largeint(self):
...
+            a = np.array([[1.1, 2], [3, 4]])
+            np.savetxt(path, a)
+            x = np.loadtxt(path)
+            assert_array_equal(x, a)
+
+    @np.testing.dec.skipif(Path is None,
+    "Cannot test without a pathlib.Path instance")
+    def test_save_load(self):
+        # Test that pathlib.Path instances can be used with savez.
+        with self.with_path(prefix='numpy_test_path_save_load_', suffix='.npy') as path:
```
Comment (charris) [inline numpy/lib/tests/test_io.py]: Umm, need to keep lines < 80 characters. This looks marginal, although can't tell on screen.

Comment (charris) [inline numpy/lib/tests/test_io.py]: could probably shorten `'numpy_test_path_save_load_'`.

**Comment by charris** [comment]
I think we are almost there. Could you squash the commits and take a look at the commit message guidelines in `doc/source/dev/gitwash/development_workflow.rst`. Not sure what to do about the context manager for temporary files, as mentioned there is another PR with one. I'd like to have one available somewhere global, maybe in `numpy/testing/utils.py`

File: numpy/lib/tests/test_io.py (hunk 8 above)
Comment (wackywendell) [inline numpy/lib/tests/test_io.py]: Ah, interesting!

I'm not sure I completely understood the problem, though.

From my understanding, a function like so:

```python
@contextlib.contextmanager
def with_file(fname):
    fd = open(fname)
    yield fd
    fd.close()
```

Will not actually run the `fd.close()` code if an exception is raised.

In this case, though, we have a `yield *inside* another with statement (`with tempdir(prefix=prefix) as tmpdir:`); AFAICT, I think that with statement (`with tempdir…`) will necessarily clean up the tempdir either way. Is that incorrect?

Comment (charris) [inline numpy/lib/tests/test_io.py]: Hmm, if we just supply a name, which I think is the right thing to do, there is nothing that needs closing, that should be handled elsewhere.

Comment (wackywendell) [inline numpy/lib/tests/test_io.py]: I don't understand. The purpose of this context manager is to 1) create a temp directory, 2) get a new temp filename in that directory, and 3) remove that directory when the context ends. What name are we supplying from where to where? How does this help?

Comment (charris) [inline numpy/lib/tests/test_io.py]: I don't really know, it needs some experimentaiton ;) Maybe aim for something like
```
with TempFilePath(...) as fpath:
```

Comment (charris) [inline numpy/lib/tests/test_io.py]: Could probably omit the directory. If I grep `mkstemp` in the repo there are no directories specified and I don't see why we would need one.

Comment (wackywendell) [inline numpy/lib/tests/test_io.py]: It's done [here](https://github.com/numpy/numpy/blob/master/numpy/lib/tests/test_io.py#L318-L325).

My understanding is that if we were to use `mkstemp`, it would only guarantee that that file would be unique *until we removed it*. For this particular test, we need to get a unique filename (done here through `mkstemp`), then *delete* that file, so that we can start with a path with nothing on it. With a bare `mkstemp`, I think its conceivable (if very unlikely) that in between closing it and reopening it the OS might give it out to another program. However, if we ask for a unique *directory*, then we don't need to worry about anything in that directory being reassigned.

Maybe its unnecessary to use both `tempdir` and `mkstemp`, but I'm not sure its 100% safe to do this without the `tempdir` if we need not just an empty file, but a valid path without a file there.

Comment (charris) [inline numpy/lib/tests/test_io.py]: Oh, forgot, we do need to delete the file (and maybe directory) when it finishes, maybe also close it just in case.

Comment (wackywendell) [inline numpy/lib/tests/test_io.py]: OK, we've now said a lot of things, and I lost track a bit of what the current problem is and/or what the solution would be.

My understanding is still that as written, the tempdir would be removed even if an exception were raised. Is that in dispute? If so, why?

Sorry for pushing the issue, its just that I'm not sure I can fix the problem if I don't understand it, because this does involve both a file that may need to be closed as well as a tempdir that needs to be removed, so with that level of complexity I'm still not sure how to proceed without a better understanding of the issue involved.

**Comment by wackywendell** [comment]
> I think we are almost there. Could you squash the commits and take a look at the commit message guidelines

Thanks! Once we figure out the context manager thing, I'll squash / rebase / fix the commit messages.

File: numpy/lib/tests/test_io.py (hunk 8 above)
Comment (charris) [inline numpy/lib/tests/test_io.py]: There are some protections, only the owner can access the file and I would expect other calls to mkstemp would have to return a different name as long as the original exists. There are also suffixes and prefixes that can be specified. The name cannot be reused by another process unless it manages to guess the name, etc, so I expect it is pretty safe. Note some of the usage patterns for mkstemp in the same test file.

For the way forward, probably the easiest thing to do is create a nice context handler in the test file and use that for this PR. Then make another PR moving the manager to `numpy/testing/utils.py` and maybe fixing up some other tests to use it.

Comment (juliantaylor) [inline numpy/lib/tests/test_io.py]: except for our tempdir context manager being broken this looks fine to me
fixing tempdir can be in a separate PR

though one could also go the route of creating a tempdir in the TestCase.setUp and delete it in tearDown, it seems each test only wants one file anyway

Comment (charris) [inline numpy/lib/tests/test_io.py]: There is also a use for temporary files in making zip files, so a context manager would be useful for more than just tests.

**Comment by rherault-insa** [comment]
+++ to @charris we definitively need context managers for temp dirs and temp files that can be closed without removed in a more larger context that tests.

**Comment by rherault-insa** [comment]
Moreover having functions accepting file descriptor and not only string or file path is a good move in the fact that they will be able to transparently accept in mem (temporary) files.

File: numpy/lib/tests/test_io.py (hunk 8 above)
Comment (charris) [inline numpy/lib/tests/test_io.py]: | then delete that file

Closing a file created by mkstemp does not delete it, that needs to be done separately.

**Comment by charris** [comment]
I've committed a `temppath` context manager in #6866. If you look there, I also refactor a test to use it. You can get at it (and tempdir) by rebasing on master and importing from `numpy.testing`.

**Comment by wackywendell** [comment]
Great! I'll look into that.

**Comment by charris** [comment]
@wackywendell Could you finish this up and squash the commits into one with a rewritten commit message ala `doc/source/dev/gitwash/development_workflow.rst`?

**Comment by wwendell** [comment]
Hi, sorry - I've been a bit side-tracked in the past month or two; I have a PhD defense in a month, and I'm going through interview stuff now, so this is taking me longer than I planned. I'll try and get it done shortly, though.

**Comment by wackywendell** [comment]
Hi, sorry - I've been a bit side-tracked in the past month or two; I have a PhD defense in a month, and I'm going through interview stuff now, so this is taking me longer than I planned. I'll try and get it done shortly, though.

**Comment by charris** [comment]
Oh, you sound busy ;) OK, I'll put this up to 1.12, but if you don't think you will have time in the near future, let me know and I'll try to finish it up.

**Comment by wackywendell** [comment]
Wow, its been a while, but I finally finished this up. It now uses the `numpy.testing.temppath` context manager for testing, and I squashed the commits. Let me know what you think!

As a side note, the pressure is off, so I should be more responsive now.

File: numpy/lib/npyio.py [hunk 11]
```diff
@@ -86,10 +86,14 @@ def __dir__(self):
         return object.__getattribute__(self, '_obj').keys()
 
 
-def zipfile_factory(*args, **kwargs):
+def zipfile_factory(file, *args, **kwargs):
+    """Create a zipfile.Zipfile. Allows for Zip64, as well as pathlib.Path
```
Comment (charris) [inline numpy/lib/npyio.py]: Please use standard docstring. Summary line, blank line, extended summary. Here that would be something like
```
"""Create a zipfile.

Extended summary.
"""
```
If you are going to document it, might say more about the arguments.

File: numpy/lib/npyio.py [hunk 12]
```diff
@@ -355,12 +359,14 @@ def load(file, mmap_mode=None, allow_pickle=True, fix_imports=True,
...
 
     """
-    import gzip
-
     own_fid = False
     if isinstance(file, basestring):
         fid = open(file, "rb")
         own_fid = True
+    elif _is_pathlib_path(file):
+        # e.g., Path instance
```
Comment (charris) [inline numpy/lib/npyio.py]: Don't need the comment, is obvious ;)

File: numpy/lib/tests/test_io.py [hunk 13]
```diff
@@ -1829,6 +1833,115 @@ def test_auto_dtype_largeint(self):
...
+    def test_loadtxt(self):
+        with temppath(suffix='.txt') as path:
+            path = Path(path)
+            a = np.array([[1.1, 2], [3, 4]])
+            np.savetxt(path, a)
+            x = np.loadtxt(path)
+            assert_array_equal(x, a)
+
+    @np.testing.dec.skipif(Path is None,
+                           "Cannot test without a pathlib.Path instance")
```
Comment (charris) [inline numpy/lib/tests/test_io.py]: Could shorten "Cannot test without a pathlib.Path instance" to "No pathlib.Path". Also below.

**Comment by charris** [comment]
Couple of nitpicky comments. I wonder if we should also deal with `memmap`? See `numpy/core/memmap.py`.

**Comment by wackywendell** [comment]
Good comments, thanks for the help.

I started work on `memmap` — that seems like another obvious place for this, thanks — but ran into an issue: what do I do about importing `numpy/lib/_iotools/_is_pathlib_path`? I can't import from `lib` into `core`, right? So I could either duplicate the functionality in `core`, or move it to `core` and then also import it into `_iotools`, which seems more right. In that case, where should it go? Maybe in `numpy.compat.py3k`, since `pathlib` is only officially in Python 3?

**Comment by wackywendell** [comment]
OK, I added another commit for `memmap`, and moved `_is_pathlib_path` to `numpy.compat.py3k`. Let me know what you think and we can roll that back or squash it.

**Comment by njsmith** [comment]
NB that there's a discussion going on on python-ideas right now about what the best approach is to making code compatible with pathlib -- some options under consideration include getting rid of pathlib (!), defining a standard protocol where every function that takes a path does `getattr(obj, "__path__", obj)` without checking the type, making pathlib Paths a subclass of `str` so that no checking is needed, etc.

**Comment by charris** [comment]
@njsmith Yeah, I noticed that pathlib was provisional in Python 3.4, makes one a bit nervous...
@wackywendell I probably would have just duplicated the function, it's not big ;)

**Comment by wackywendell** [comment]
@charris: OK, do you think I should just duplicate the function? Or is it OK to put it `compat.py3k`?

**Comment by charris** [comment]
@wackywendell Not worth worrying about, you can leave it as is.

**Comment by charris** [comment]
Although it might be worth removing the leading `_` as it isn't remotely private anymore.

**Comment by wackywendell** [comment]
OK, done, done, and rebased! Anything else?

**Comment by njsmith** [comment]
Only concern is that it sounds like python-dev is converging on a nicer way to do this that will probably be released in the next 3.4, 3.5 point releases. Maybe we don't care and can always clean this up later once that happens -- I don't have a strong opinion either way.

**Comment by njsmith** [comment]
Probably we should just go for it given that the PR exists :-). I won't merge myself though b/c I'm just thinking about the high-level issue, I haven't actually read the code.

File: numpy/compat/py3k.py [hunk 14]
```diff
@@ -10,6 +10,10 @@
            'integer_types']
```
Comment (charris) [inline numpy/compat/py3k.py]: Add `Path` and `is_pathlib_path` to the `__all__` list, then you can import from `numpy.compat`.

File: numpy/core/memmap.py [hunk 15]
```diff
@@ -3,6 +3,7 @@
 import numpy as np
 from .numeric import uint8, ndarray, dtype
 from numpy.compat import long, basestring
+from numpy.compat.py3k import is_pathlib_path
```
Comment (charris) [inline numpy/core/memmap.py]: Should be able to import directly from `numpy.compat` after fix.

File: numpy/core/tests/test_memmap.py [hunk 16]
```diff
@@ -4,6 +4,10 @@
 import os
 import shutil
 from tempfile import NamedTemporaryFile, TemporaryFile, mktemp, mkdtemp
+try:
```
Comment (charris) [inline numpy/core/tests/test_memmap.py]: Could import `Path` from `npy_compat`

File: numpy/lib/_iotools.py [hunk 17]
```diff
@@ -9,6 +9,7 @@
 import numpy as np
 import numpy.core.numeric as nx
 from numpy.compat import asbytes, bytes, asbytes_nested, basestring
+from numpy.compat.py3k import is_pathlib_path
```
Comment (charris) [inline numpy/lib/_iotools.py]: Best import this from `numpy.compat` so everything is in one place.

Comment (charris) [inline numpy/lib/_iotools.py]: In fact, I don't think you need this anymore.

File: numpy/lib/npyio.py [hunk 18]
```diff
@@ -14,8 +14,8 @@
 from numpy.core.multiarray import packbits, unpackbits
 from ._iotools import (
     LineSplitter, NameValidator, StringConverter, ConverterError,
-    ConverterLockError, ConversionWarning, _is_string_like, has_nested_fields,
-    flatten_dtype, easy_dtype, _bytes_to_name
+    ConverterLockError, ConversionWarning, _is_string_like, is_pathlib_path,
```
Comment (charris) [inline numpy/lib/npyio.py]: Might as well import from `npy_compat` so that everything is in one place.

File: numpy/lib/npyio.py [hunk 19]
```diff
@@ -86,10 +86,17 @@ def __dir__(self):
         return object.__getattribute__(self, '_obj').keys()
 
 
-def zipfile_factory(*args, **kwargs):
+def zipfile_factory(file, *args, **kwargs):
```
Comment (charris) [inline numpy/lib/npyio.py]: Should be OK, looks like zipfile_factory is effectively private.

**Comment by charris** [comment]
I think we are getting close. @njsmith Should be easy to search and fixup. What changes are being proposed?

File: numpy/lib/npyio.py (hunk 19 above)
Comment (wackywendell) [inline numpy/lib/npyio.py]: This is not changing the effective signature - `*args` was being passed to `zipfile.ZipFile` before, and `zipfile.ZipFile` takes `file` as its first argument, so anything that worked before should continue to work.

**Comment by njsmith** [comment]
@charris: it looks like the current trend is towards adding a standard convert-to-filesystem-path protocol sorta along the lines of `__index__`, so APIs that take paths and want strings would do `from operators import fspath` and then
```python
def some_file_function(path, ...):
    # force path to be a str
    path = fspath(path)
    ...
```
Or something like that.

Also probably pathlib support will get added to a bunch of builtins like `open`, so that for most code no special handling will be needed at all -- you just write
```
def some_file_function(path, ...):
    with open(path) as f:
        ...
```
and it just magically works for both str and Path arguments.

**Comment by njsmith** [comment]
But, it's a mailing list discussion, so -- that's where things seem to be going right now, but, check back next week for our thrilling conclusion :-).

File: numpy/lib/tests/test_io.py [hunk 20]
```diff
@@ -1829,6 +1833,115 @@ def test_auto_dtype_largeint(self):
         assert_equal(test['f1'], 17179869184)
         assert_equal(test['f2'], 1024)
 
+
+class TestPathUsage(TestCase):
+    # Test that pathlib.Path can be used
+    @np.testing.dec.skipif(Path is None,
```
Comment (charris) [inline numpy/lib/tests/test_io.py]: Umm, this will now fit on a single line. Same below.

File: numpy/lib/npyio.py [hunk 21]
```diff
@@ -86,10 +86,17 @@ def __dir__(self):
...
+    """Create a ZipFile.
+    
+    Allows for Zip64, and the `file` argument can accept file, str, or
+    pathlib.Path objects. `args` and `kwargs` are passed to the zipfile.ZipFile
+    constructor."""
```
Comment (charris) [inline numpy/lib/npyio.py]: Put a blank line and then the `"""` aligned with the first.

**Comment by charris** [comment]
Looks good modulo two style nits.

**Comment by charris** [comment]
Thanks @wackywendell .

**Comment by ChrisBarker-NOAA** [comment]
Sorry to not look at the code -- but is this going to work in py2 with the pypi pathlib2 object? If so -- great!

**Comment by charris** [comment]
No, it won't work with `pathlib2` because the module name differs. @wackywendell That could probably be fixed.

**Comment by ChrisBarker-NOAA** [comment]
well, there is a new protocol being proposed -- something like a __fspath__ dunder method. So maybe just wait 'till that gets finalized.

**Comment by charris** [comment]
@ChrisBarker-NOAA Probably a good idea. There is also `tofile` and `fromfile` to worry about, but they are at the C level and so it will be worth waiting to see what the final interface is.

**Comment by eric-wieser** [comment]
That protocol has been finalized in [PEP 519](https://www.python.org/dev/peps/pep-0519/), and is live in python 3.6
